[CIR] Better handling of `void` function return

dkolsen-pgi opened the pull request:
Upstream https://github.com/llvm/clangir/pull/1249 "Remove return !cir.void from IR and textual representation"

C/C++ functions returning `void` had an explicit `!cir.void` return type while not having any returned value, which was breaking MLIR invariants when the CIR dialect is used in a greater context, for example with the inliner.

Now a C/C++ function returning `void` has no return type and no return values, which satisfies the MLIR invariant about the same number of return types and returned values.

dkolsen-pgi committed:
[CIR] Better handling of `void` function return

Upstream https://github.com/llvm/clangir/pull/1249
"Remove return !cir.void from IR and textual representation"

C/C++ functions returning `void` had an explicit `!cir.void` return type
while not having any returned value, which was breaking MLIR invariants
when the CIR dialect is used in a greater context, for example with the
inliner.

Now a C/C++ function returning `void` has no return type and no return
values, which satisfies the MLIR invariant about the same number of
return types and returned values.

dkolsen-pgi commented:
@keryell Please review this, since it is your change.  (I don't know why GitHub won't let me add you as a reviewer.)

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -273,29 +273,36 @@ def CIR_PointerType : CIR_Type<"Pointer", "ptr",
   }];
 
   let builders = [
+    // Construct with an actual return type or explicit !cir.void
```
Should this be 'implicit' void?  It actually looks like this is just doing 'return type or nothing (if void)'.  

Also, clang comments should end with a full-stop.

dkolsen-pgi replied:
I agree.  I am changing "explicit" to "implicit".

keryell replied:
No, I really meant explicit here.
If you put an explicit `!cir.void` in the assembly text, it will be discarded and will be nowhere in the IR.
It is as if there was
This is for compatibility with pas ClangIR behavior.
Of course, since we are up-streaming it, do we need this backward compatibility? If not, it might breaks more stuff while rebasing ClangIR downstream... :thinking:

bcardosolopes replied:
Even in the incubator it should be fine to remove the compatibility given that I don't know any users loading/emitting old CIR

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -309,11 +316,15 @@ def CIR_FuncType : CIR_Type<"Func", "func"> {
     unsigned getNumInputs() const { return getInputs().size(); }
 
+    /// Returns the result type of the function as an actual return type or
+    /// explicit !cir.void
```
Same comments here.

dkolsen-pgi replied:
In this case, "explicit" is correct.  The definition of `getReturnType` returns a `!cir.void` if the MLIR type has no return type.  This function should be called when looking at things from a C++ point of view instead of an MLIR point of view.

erichkeane commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -331,9 +335,38 @@ FuncType FuncType::clone(TypeRange inputs, TypeRange results) const {
+                                mlir::Type optionalReturnType) {
+  if (optionalReturnType)
+    p << optionalReturnType << ' ';
+  p << '(';
```
It seems a little weird that THIS function is the one that prints the open paren?  Shouldn't whoever calls this print it?

dkolsen-pgi replied:
One could argue that since `parseFuncTypeReturn` parses the opening `(` of the parameter list, `printFuncTypeReturn` should match and print the opening `(` of the parameter list.

But I'm not going to argue that.  I agree with you.  It makes more sense if the `(` is output as part of `printFuncTypeArgs` instead of `printFuncTypeReturn`, even if the parse functions do it differently.  I'll make that change.

erichkeane commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -331,9 +335,38 @@ FuncType FuncType::clone(TypeRange inputs, TypeRange results) const {
+static mlir::ParseResult parseFuncTypeReturn(mlir::AsmParser &p,
+                                             mlir::Type &optionalReturnType) {
+  if (succeeded(p.parseOptionalLParen())) {
+    // If we have already a '(', the function has no return type
```
I'm perhaps a little confused as to what the format here is we're trying to parse.  Why is just a `(` mean `empty return type`?  

Or is this supposed to be the 'rest' of the function type?  In that case, perhaps we need a 'look-ahead' instead of a parse?

keryell replied:
Yes this is a kind of look-ahead which is required when I introduced to remove the creation of an artificial `!cir.void` which had the consequence of breaking an MLIR invariant, `number(return-types) == number(return-values)`.
A C type like `char(int)` is lowered and pretty-printed as `!cir.func<!cir.int<s, 8>(!cir.int<s, 32>)` while `void(int)` is lowered and pretty-printed as `!cir.func<(!cir.int<s, 32>)`.
For the MLIR functions themselves, they are handled by the `func` MLIR standard dialect with a syntax like:
```mlir
func.func @count(%x: i64) -> (i64, i64)
// The same returning nothing:
func.func @f(%x: i64)
```
with an empty type for returning "void".
An alternate design could be to have a new keyword like `void` understood by the parser to avoid the manual look ahead in the parser as the MLIR default parser knows how to discriminate with prefix keywords. Then `void(int)` would be lowered and pretty-printed as `!cir.func<void (!cir.int<s, 32>)`.

erichkeane commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -331,9 +335,38 @@ FuncType FuncType::clone(TypeRange inputs, TypeRange results) const {
+  if (p.parseType(type))
+    return mlir::failure();
+  if (isa<cir::VoidType>(type))
+    // An explicit !cir.void means also no return type.
```
Typically a comment is enough to cause us to not do the `omit braces` rule here, so I suggest putting curleys around this if and else.

erichkeane commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -331,9 +335,38 @@ FuncType FuncType::clone(TypeRange inputs, TypeRange results) const {
+                  bool &isVarArg) {
   isVarArg = false;
   // `(` `)`
   if (succeeded(p.parseOptionalRParen()))
```
I see what is happening here now, not sure I like it?  IMO, the parse-args function should be responsible for 'taking' BOTH of the parens, everything else should do a 'look-ahead' type parsing if necessary.  THOUGH I see the latter doesn't seem to exist in `mlir::AsmParser`?  Is this a case of us 'using it wrong'? What do other parsers do here?

dkolsen-pgi replied:
I am not familiar with `mlir::AsmParser`, but looking through its definition just now I don't see anyway to peek at the next token or do lookahead or do a trial parse without advancing.  I agree that `parseFuncTypeArgs` should handle both the `(` and the `)`.  But I don't see a way to do that.  I don't have experience with other MLIR parsers, so I don't know what they do.  Unless someone more knowledgeable pipes up, I don't see a way to change this.

erichkeane replied:
@lanza @bcardosolopes Anyone have any ideas here?  This parsing is REALLY weird, but we still need a way to know that we're 'done'.

The other alternative is to change how these are printed/read (so it is something like: `(returnType)(Arg1, Arg2)` with whatever delimiters make sense), but I'm not sure that is something that can be done sensibly.

keryell replied:
I think this is the way the parsers are implemented to chain things ahead.
See again my comment with `void` keyword.
At the end it is a trade-off between implementator comfort and code beauty vs the concrete assembly syntax exposed to the end user.

xlauko replied:
You can technically do it the other way around as mlir function types are structured and have `!cir.func<(args) -> ret>` or `!cir.func<(args)>` in case of void. This can be parsed without lookahead using `parseOptionalArrow`.

xlauko replied:
`(args)` can be nicely parsed using `parseCommaSeparatedList(mlir::AsmParser::Delimiter::Paren, parseArg)` which takes care of both `parens`

andykaylor replied:
I like the idea of the `!cir.func<(args) -> ret>` format. It feels more consistent with func dialect.

keryell replied:
I like this idea too.
I have tried to be as close as the original syntax to avoid disrupting too much the tests and the users but there is no reason to do something else.
If there is a consensus from the ClangIR stake-holders, go for it!

xlauko replied:
@bcardosolopes @lanza ^

erichkeane replied:
The only question is what the 'void' type looks like?  Could/should we still use the `void` as the return type in the textual representation, or alternatively just omit entirely? So:

`!cir.func<(args) -> void>`
vs: 
`!cir.func<(args)>`

That way we do `parse arguments`, then `if arrow, parse return type, else is void`.

The 'trailing return type' syntax is nice as it is consistent with the C++ language.

erichkeane replied:
Got it, thanks for the update!  I was trying to judge opinions here, but the incubator review is sufficient to make sure those opinions are heard.

bcardosolopes replied:
> I like the idea of the !cir.func<(args) -> ret> format

+1

dkolsen-pgi committed:
Followup to void return functions in CIR

Fix some comments and some formatting.  Change which function prints the
opening `(` of the parameter list of a function type.  No changes in
behavior.

xlauko commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -375,11 +409,48 @@ void printFuncTypeArgs(mlir::AsmPrinter &p, mlir::ArrayRef<mlir::Type> params,
+mlir::Type FuncType::getReturnType() const {
+  if (isVoid())
+    return cir::VoidType::get(getContext());
+  return static_cast<detail::FuncTypeStorage *>(getImpl())->optionalReturnType;
```
```suggestion
  return getOptionalReturnType();
```

xlauko commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -375,11 +409,48 @@ void printFuncTypeArgs(mlir::AsmPrinter &p, mlir::ArrayRef<mlir::Type> params,
-  return static_cast<detail::FuncTypeStorage *>(getImpl())->returnType;
+  if (isVoid())
+    return {};
+  return static_cast<detail::FuncTypeStorage *>(getImpl())->optionalReturnType;
```
```suggestion
  return getOptionalReturnType();
```

xlauko replied:
Also to mirror `getReturnType` shouldn't this return `cir::VoidType::get(getContext())` in void case?

keryell replied:
No because it returns the real return types (plural) that any real MLIR user is waiting for: the actual return types, which in the case of C/C++ is 0 or 1 type.
The whole point of this PR is to fix this bug and not to return some synthetic `!cir.void` which was breaking the neighborhood.

xlauko commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -375,11 +409,48 @@ void printFuncTypeArgs(mlir::AsmPrinter &p, mlir::ArrayRef<mlir::Type> params,
+// Whether the function returns void
+bool FuncType::isVoid() const {
+  auto rt =
+      static_cast<detail::FuncTypeStorage *>(getImpl())->optionalReturnType;
```
```suggestion
      getOptionalReturnType();
```

xlauko commented on clang/test/CIR/func-simple.cpp:
```diff
@@ -2,12 +2,12 @@
 // RUN: %clang_cc1 -std=c++20 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o -  | FileCheck %s
```
I presume `clang` cannot load `cir` files yet? So nothing is really testing parsers?

andykaylor replied:
It's definitely unfortunate that we're adding this parsing code but don't have any tests for it. I'm working on a PR to bring cir-opt over so we can do that.

andykaylor replied:
https://github.com/llvm/llvm-project/pull/128254

xlauko commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -375,11 +410,48 @@ void printFuncTypeArgs(mlir::AsmPrinter &p, mlir::ArrayRef<mlir::Type> params,
+// Use a custom parser to handle the optional return and argument types without
+// an optional anchor.
+static mlir::ParseResult parseFuncType(mlir::AsmParser &p,
+                                       mlir::Type &optionalReturnTypes,
```
```suggestion
                                       mlir::Type &optionalReturnType,
```

xlauko commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -375,11 +410,48 @@ void printFuncTypeArgs(mlir::AsmPrinter &p, mlir::ArrayRef<mlir::Type> params,
+  return parseFuncTypeArgs(p, params, isVarArg);
+}
+
+static void printFuncType(mlir::AsmPrinter &p, mlir::Type optionalReturnTypes,
```
```suggestion
static void printFuncType(mlir::AsmPrinter &p, mlir::Type optionalReturnType,
```

xlauko commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -375,11 +410,48 @@ void printFuncTypeArgs(mlir::AsmPrinter &p, mlir::ArrayRef<mlir::Type> params,
+                                       mlir::Type &optionalReturnTypes,
+                                       llvm::SmallVector<mlir::Type> &params,
+                                       bool &isVarArg) {
+  if (failed(parseFuncTypeReturn(p, optionalReturnTypes)))
```
```suggestion
  if (failed(parseFuncTypeReturn(p, optionalReturnType)))
```

xlauko commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -375,11 +410,48 @@ void printFuncTypeArgs(mlir::AsmPrinter &p, mlir::ArrayRef<mlir::Type> params,
+
+static void printFuncType(mlir::AsmPrinter &p, mlir::Type optionalReturnTypes,
+                          mlir::ArrayRef<mlir::Type> params, bool isVarArg) {
+  printFuncTypeReturn(p, optionalReturnTypes);
```
```suggestion
  printFuncTypeReturn(p, optionalReturnType);
```

keryell reviewed (commented):
Sorry for being late to the party!

dkolsen-pgi commented:
I like the idea of changing the assembly format for function types from `!cir.func<!returnType (!argType)>` to `!cir.func<(!argType) -> !returnType>`.  That is
1. Easier to parse.
2. Consistent with function types in other MLIR dialects.
3. Consistent with the assembly format for function definitions in ClangIR.

Making a change like this really needs to happen in the incubator first, not as part of upstreaming.  So my current plan is to

1. abandon this PR,
2. create a new PR in the incubator to make this change (including fixing lots of tests),
3. and then create a new upstream PR with the new behavior.

bcardosolopes commented:
Thanks for the update @dkolsen-pgi, reviewing the incubator change today

dkolsen-pgi commented:
Abandoning this PR.  I will redo it soon with the new assembly format for function types.
